packages/e2e-tests/specs/editor/various/rich-text.test.js

```diff
@@ -426,30 +452,47 @@ describe( 'RichText', () => {
 		await page.keyboard.press( 'ArrowLeft' );
 		await page.keyboard.type( '1' );
 
-		// Expect '1🍓'.
+		// Expect the text to be typed correctly.
+		await page.waitForSelector( '1🍓', { state: 'visible' } );
 		expect( await getEditedPostContent() ).toMatchSnapshot();
 	} );
 
 	it( 'should show/hide toolbar when entering/exiting format', async () => {
 		const blockToolbarSelector = '.block-editor-block-toolbar';
 		await clickBlockAppender();
+
 		await page.keyboard.type( '1' );
-		expect( await page.$( blockToolbarSelector ) ).toBe( null );
+		await page.waitForSelector( blockToolbarSelector, { state: 'hidden' } );
+
 		await pressKeyWithModifier( 'primary', 'b' );
-		expect( await page.$( blockToolbarSelector ) ).not.toBe( null );
+		await page.waitForSelector( blockToolbarSelector, {
+			state: 'visible',
+		} );
+
 		await page.keyboard.type( '2' );
-		expect( await page.$( blockToolbarSelector ) ).not.toBe( null );
+		await page.waitForSelector( blockToolbarSelector, {
+			state: 'visible',
+		} );
+
 		await pressKeyWithModifier( 'primary', 'b' );
-		expect( await page.$( blockToolbarSelector ) ).toBe( null );
+		await page.waitForSelector( blockToolbarSelector, { state: 'hidden' } );
+
 		await page.keyboard.type( '3' );
 		await page.keyboard.press( 'ArrowLeft' );
-		expect( await page.$( blockToolbarSelector ) ).toBe( null );
+		await page.waitForSelector( blockToolbarSelector, { state: 'hidden' } );
+
 		await page.keyboard.press( 'ArrowLeft' );
-		expect( await page.$( blockToolbarSelector ) ).not.toBe( null );
+		await page.waitForSelector( blockToolbarSelector, {
+			state: 'visible',
+		} );
+
 		await page.keyboard.press( 'ArrowLeft' );
-		expect( await page.$( blockToolbarSelector ) ).not.toBe( null );
+		await page.waitForSelector( blockToolbarSelector, {
+			state: 'visible',
+		} );
+
 		await page.keyboard.press( 'ArrowLeft' );
-		expect( await page.$( blockToolbarSelector ) ).toBe( null );
```

Comment:
It's also no the same here that the previous statement expects the node to be not found as soon as possible, but the current one expects it to be hidden in a certain amount of time. Ideally, the current one is almost always better though, but I don't know much about the specific test here.

Reply:
> It's not clear that awaiting for selectors is an adequate substitute for the assertions, as these provide us with easy-to-read error messages in case of failure.

They're not adequate, but in general waiting for elements should not add any extra time, and at the same time should be less prone to flakiness. Also, in this particular case, IMO this:

```shell
elementHandle.waitForElementState: Timeout 3000ms exceeded.
=========================== logs ===========================
  waiting for element to be hidden
    element is not hidden - waiting...
============================================================
```

is easier to follow than this:

```shell
expect(received).toBe(expected) // Object.is equality

Expected: null
Received: {}
```

---

> It's also no the same here that the previous statement expects the node to be not found as soon as possible, but the current one expects it to be hidden in a certain amount of time.

As mentioned above, in general, awaiting elements/state is just less prone to flakiness as often DOM changes don't happen immediately. In my testing with Playwright, this is especially the case when performing editor actions, like pressing `Enter` to create a new paragraph, or `Backspace` to remove a block. When the keypress-followed actions were expected immediately, they were often producing flaky output.

Reply:
Yeah, I mean that I agree this change is better 😅.

Reply:
It's also worth mentioning here that Playwright's dedicated test-runner uses `expect` assertions with all the auto-waiting goodness: https://playwright.dev/docs/test-assertions.